### ISSUE TITLE
Enhance Compatibility debugging workflow narrative

### ISSUE THREAD

**Issue by amedina** [ampproject/amp-wp#2023]
The AMP plugin functionality is centered around interactions between `actions taken by the plugin` and `actions taken by the user`. The plugin can be configured to detect validation errors and automatically remove them (automatic sanitization). Or it can be configured to only detect errors but not removing them (manual sanitization). The user/developer has the control of what to do with detected errors. He/she can `accept` the plugin action (remove/not remove), or he/she can `reject` the plugin action. 

This dicothomy between plugin and user actions percolates to the compatibility tool, and it is a main source of confusion for users of the plugin. This manifests in the labeling of errors that are surfaced by the compatibility tool as `New Accepted` or `New Rejected`, and the labeling of errors after the user has taken an action: `Accepted` or `Rejected`. 

The problem is that the plugin actions are `Remove` and `Keep` (referring to the fate of the offending markup causing an error), while the user action is `Accept` or `Reject` (referring to the user intention with respect to the plugin action).

To remove this confusion, the compatibility tool should label the errors according to the action of the plugin, and not according to the action of the user. This entails updating the error labels this way:

- `New Accepted` to `New, Removed`
- `New Rejected` to `New, Kept`
- `Accepted` to `Removed`
- `Rejected` to `Kept`

This is only a change in labels and internal resource labeling.

**Comment by westonruter** [comment]
See also https://github.com/ampproject/amp-wp/issues/1741

Does this terminology change work for "Statuses"? 

Since the underlying entity is the "validation error", does it make sense that a "validation error is removed" and a "validation error is kept"? Or does removed/kept make more sense if the underlying entity is the "invalid markup"?

**Comment by amedina** [comment]
I see that point. Still we are on the right track. Let's flesh out things a little more. The ultimate goal should be two fold: 

1. Decouple `plugin actions` from `user actions`
2. Maintain consistency on the plugin side so that labels and statuses refer to either plugin actions or error-specific labels.

The status of an validation error as `removed` captures the fact that the error is not present in the generated content. That makes sense for status labels: this is an error which has been removed from the content. Similarly, the status of an error can be `Kept`, meaning it is still present in the generated content. 

There are no other statuses for errors; that is, there is not a 'fixed' status, or anything else. And this approach satisfies conditions (1) and (2) above. 

WDYT?

**Comment by westonruter** [comment]
As discussed, it may be better to speak of “Retained” rather than “Kept”.

Something else to consider is that if:

* ❌ Rejected ➡️ Retained/Kept
* ✅ Accepted ➡️ Removed

Do these icons still make sense?

**Comment by westonruter** [comment]
For master issue on the Compatibility Tool redesign, see #2316.

**Comment by amedina** [comment]
Discussing today with @pbakaus we considered using: 

❌ Rejected ➡️ Markup Retained
✅ Accepted ➡️ Markup Removed

Which aligns well with the sanitization actions by the plugin.

**Comment by westonruter** [comment]
The only gotcha with “Markup” is that the validation error could be due to non-markup code, like an illegal CSS at-rule or excessive CSS.

**Comment by pbakaus** [comment]
Retained is a pretty complicated word, that many non-native speakers will not understand, I'm afraid.

How about:
-> Remove invalid content
-> Keep invalid content

We could also replace invalid with incompatible, might be clearer.

**Comment by amedina** [comment]
The label must capture the status of the markup (past tense). We can have:

❌ Rejected ➡️ Markup Kept
✅ Accepted ➡️ Markup Removed

**Comment by amedina** [comment]
Agree with using content instead of markup for generality and simplicity

**Comment by westonruter** [comment]
What about “Code” instead of “Content”?

**Comment by postphotos** [comment]
YES! This is a major step in the right direction. 🚀

Calling out Weston's note above, sometimes it's an illegal CSS rule or excessive that is flagging the sanitizer, not markup.

So if it's helpful, I'd suggest a slight adjustment, but with the same line of thinking:
❌ `Rejected` ➡️ `Content Kept (AMP Not Served)`
✅ `Accepted` ➡️ `Incompatible Content Hidden`

Maybe that's too wordy, but at least it's well understood.

**Comment by westonruter** [comment]
Reject: 🗑
Accept: 💾

😉

**Comment by westonruter** [comment]
Another alternative to “accept” is “suppress”, as the acceptance is intended to be temporary. The underlying issue should ideally be fixed in the theme/plugin that caused the problem in the first place.

**Comment by westonruter** [comment]
Stepping back from this to get a fresh look.

----

The AMP plugin has an internalized representation of the AMP validator specification. The internalized validation rules are used in the plugin's sanitizers to prevent invalid AMP markup from being served on the frontend (to prevent Google Search Console from complaining). At the same time, the AMP plugin provides warnings about those caught validation errors in order to ensure the user is aware that it performing any sanitization actions. 

Validation errors that have already been seen need to be marked as acknowledged in order to differentiate them from newly-encountered validation errors that require the user to review. 

The user also needs to be able to opt-out of the markup causing a validation error from being sanitized. It may be that some invalid AMP markup on the page is actual critical for the page to function properly (e.g. some interactive graphic), and in this case the user needs to reject it from being sanitized. When a sanitization for a validation error's markup is rejected, there are two outcomes:

* On a Transitional mode site (which is paired AMP), the user is redirected from the AMP version to the non-AMP version, and the rel=amphtml link is not even presented for crawling.
* On a Standard mode site (which is AMP-first), there are no non-AMP URLs and thus redirection is not an option. In this case the amp attribute is removed from the root HTML element, preventing the page from being validated as an AMP page (and thus preventing Google Search Console from complaining). Such pages still use the AMP framework, and eventually these would be utilizing Bento AMP.

As such, a validation error which is sanitized is shown with ✅ whereas a validation error which is not sanitized is shown with ❌, since whether something is sanitized determines whether AMP is available.

There are 4 possible states for a validation error:

* ✅ New Accepted 
* ✅ (Ack) Accepted 
* ❌ New Rejected 
* ❌ (Ack) rejected 

Again, the “new” states are important because they highlight validation errors which haven't been dealt with yet. So the “new” here is synonymous with “unacknowledged”, “unread” or “unapproved” (which is used on the Comments screen):

<img width="1422" alt="Screen Shot 2019-10-25 at 12 04 55" src="https://user-images.githubusercontent.com/134745/67597299-b1c47580-f71f-11e9-89b2-9d3715b7c676.png">
 
On the Validated URL screen, here is what it currently looks like when a URL has validation errors in all 4 possible states:

<img width="1429" alt="Screen Shot 2019-10-25 at 11 30 32" src="https://user-images.githubusercontent.com/134745/67595206-ebdf4880-f71a-11e9-9b7a-9309b946a37b.png">

Here is what it could look like if we replace the “new” with the concept of read/unread state, and rebrand “accepted” as “removed” and “rejected” as “kept” (while also rebranding “Status” to “Action”):

<img width="1423" alt="Screen Shot 2019-10-25 at 11 28 32" src="https://user-images.githubusercontent.com/134745/67595329-3d87d300-f71b-11e9-86e0-00d057aa33fa.png">

<img width="1427" alt="Screen Shot 2019-10-25 at 11 28 49" src="https://user-images.githubusercontent.com/134745/67595330-3d87d300-f71b-11e9-8bd5-0ba615fcc44a.png">

The semantics change here as instead of listing validation errors it is rather listing invalid markup which is _causing_ validation errors. So instead of manipulating the status of whether or not a validation error should (have its underlying markup) be sanitized, it changes to what action should be performed with the invalid AMP markup.

Notice that there is now no longer any “New” in the dropdowns. This is moved to a “Mark as read” row action, which would behave similar to the “Approve” link for comments. It doesn't change the action, but just clears out the unmoderated state. There is also here a new capability, to mark a previously read error as _unread_ (via the “Mark as unread“ row action link. This would provide a way for a user to essentially flag a validation error to revisit later, similar to what people do with marking emails as unread. 

Clicking “mark as read” and “mark as unread” would not cause the page to reload (same as comments) and the user wouldn't have to click the Update to persist the changes. The only concern I have with this is the call to action for clearing out the “unread” is not as easy to find as perhaps should be. Maybe instead of “read”/“unread” there should be instead (following the email paradigm) be the concept of “starring”. There could be a new column with a star which the user could click to toggle the state, which would make this more discoverable, for example:

<img width="1420" alt="Screen Shot 2019-10-25 at 11 49 53" src="https://user-images.githubusercontent.com/134745/67596465-9c4e4c00-f71d-11e9-94a4-5f0ccaf260c1.png">

Or using unseen/seen instead of unstarred/starred:

<img width="1427" alt="Screen Shot 2019-10-25 at 11 47 32" src="https://user-images.githubusercontent.com/134745/67596306-45e10d80-f71d-11e9-9c70-a1c2c1508c78.png">

Or going with a checkbox:

<img width="1420" alt="Screen Shot 2019-10-25 at 11 52 32" src="https://user-images.githubusercontent.com/134745/67596610-f94a0200-f71d-11e9-9180-d1830bac7592.png">

But I don't like this so much because there is already a checkbox in the first column for bulk actions, and checkboxes don't usually change state when clicked (which we'd be doing here), and it confuses with the other ✅/❌ iconography (which perhaps should also be revamped).

**Comment by westonruter** [comment]
Another alternative to seen/unseen is reviewed/unreviewed.

**Comment by westonruter** [comment]
## Current Validated URL Screens

As for the list of Validated URLs, here is what it looks like presently:

<img width="1425" alt="Screen Shot 2019-10-25 at 12 30 54" src="https://user-images.githubusercontent.com/134745/67598875-817ed600-f723-11e9-9df8-69695a4c859d.png">
<img width="1422" alt="Screen Shot 2019-10-25 at 12 31 15" src="https://user-images.githubusercontent.com/134745/67598876-82176c80-f723-11e9-9890-e63c882e95f2.png">
<img width="1431" alt="Screen Shot 2019-10-25 at 12 31 27" src="https://user-images.githubusercontent.com/134745/67598877-82176c80-f723-11e9-8125-a1ba683711a4.png">

The “new” state for the URL shown here is present because it has one or more validation errors that are in this new/unseen/unacknowledged/unapproved state.

## Mocks for Updated Validated URL Screens

<img width="1429" alt="Screen Shot 2019-10-25 at 12 33 57" src="https://user-images.githubusercontent.com/134745/67599012-d91d4180-f723-11e9-808d-3630982ffeec.png">
<img width="1431" alt="Screen Shot 2019-10-25 at 12 34 09" src="https://user-images.githubusercontent.com/134745/67599015-d91d4180-f723-11e9-9177-b1704f2f8f91.png">
<img width="1427" alt="Screen Shot 2019-10-25 at 12 34 18" src="https://user-images.githubusercontent.com/134745/67599016-d91d4180-f723-11e9-9ca0-e522ebe34df9.png">

Something that is missing from this view is a column indicating whether AMP is enabled/disabled for the URL. This is implied by there being a validation error in the “Kept” state, but it could be more clear.

**Comment by westonruter** [comment]
## Current Validation Error Index Screen

<img width="1425" alt="Screen Shot 2019-10-25 at 12 39 18" src="https://user-images.githubusercontent.com/134745/67599284-960f9e00-f724-11e9-8c9a-de6a60298488.png">
<img width="1428" alt="Screen Shot 2019-10-25 at 12 39 28" src="https://user-images.githubusercontent.com/134745/67599285-960f9e00-f724-11e9-8184-ea0321d07656.png">
<img width="1430" alt="Screen Shot 2019-10-25 at 12 39 51" src="https://user-images.githubusercontent.com/134745/67599286-960f9e00-f724-11e9-860f-b98db5229c7f.png">
<img width="1434" alt="Screen Shot 2019-10-25 at 12 39 58" src="https://user-images.githubusercontent.com/134745/67599287-96a83480-f724-11e9-901e-f0e72aa80047.png">

# Mocks for Updated Validated URL Screen

See discussion above about whether stars, eyeballs, or something else should be used to represent the acknowledged state.

<img width="1425" alt="Screen Shot 2019-10-25 at 12 46 19" src="https://user-images.githubusercontent.com/134745/67599734-86dd2000-f725-11e9-9280-9888d7f97136.png">
<img width="1431" alt="Screen Shot 2019-10-25 at 12 46 28" src="https://user-images.githubusercontent.com/134745/67599735-86dd2000-f725-11e9-9e73-bfe00384c286.png">
<img width="1428" alt="Screen Shot 2019-10-25 at 12 46 37" src="https://user-images.githubusercontent.com/134745/67599738-86dd2000-f725-11e9-99c3-dadf9f48b2f6.png">

**Comment by westonruter** [comment]
As an alternative to unseen/starred/unapproved, another workable concept would be **flagged**.

A validation error that not been reviewed yet, could get flagged by default, indicating that it is an issue that needs to be reviewed. This concept would also work for issues which actually have been seen before but which the user wants to put back into an “unread” state. There is a [flag Dashicon](https://developer.wordpress.org/resource/dashicons/#flag):

![image](https://user-images.githubusercontent.com/134745/67600040-28647180-f726-11e9-9d95-853802df2c87.png)

But there is no corresponding “unflagged” dashicon similar to there is for filled/unfilled star or visible/hidden:

![image](https://user-images.githubusercontent.com/134745/67600166-5649b600-f726-11e9-8d44-b8a25785e83f.png) ![image](https://user-images.githubusercontent.com/134745/67600175-5ea1f100-f726-11e9-8763-a006d2c4d24e.png)

![image](https://user-images.githubusercontent.com/134745/67600189-6a8db300-f726-11e9-9a59-86206e49f10c.png) ![image](https://user-images.githubusercontent.com/134745/67600208-74afb180-f726-11e9-96fb-304a48b910c0.png)

**Comment by westonruter** [comment]
Here's where @amedina and I have settled:

* Remove the “Seen” column entirely, in favor of the “Mark as seen” and “Mark as unseen” row actions.
* Rename the “Action” column to “Markup Status”.
* Changing the seen state of a validation error is persisted immediately upon clicking (as is done with Comments), whereas the removed/kept status is only persisted when clicking Update (when on the Validated URL screen).
* Use “Remove markup” and “Keep markup” for the action buttons and bulk action options.

<img width="1355" alt="Screen Shot 2019-10-25 at 16 52 56" src="https://user-images.githubusercontent.com/134745/67610566-84d88880-f748-11e9-87b8-22e140ac8416.png">
<img width="540" alt="Screen Shot 2019-10-25 at 16 54 17" src="https://user-images.githubusercontent.com/134745/67610567-85711f00-f748-11e9-96c6-78b4090f20dc.png">

**Comment by westonruter** [comment]
I think we need better icons to represent Removed and Kept:

![image](https://user-images.githubusercontent.com/134745/67628606-c0955000-f825-11e9-9766-d966b2ae59f2.png)

The ✅ is currently used for removal because it has a _positive_ result for AMP, in that it allows AMP to be served. The ❌ is used for keeping because it has a _negative_ result for AMP, that it blocks AMP from being served for the given URL.

Possibilities:

![image](https://user-images.githubusercontent.com/134745/67628624-0d792680-f826-11e9-99cd-529302a55afe.png)
<img  alt="Screen Shot 2019-10-26 at 19 23 41" src="https://user-images.githubusercontent.com/134745/67628641-3b5e6b00-f826-11e9-9a0b-6c377d37bfd5.png"> ![image](https://user-images.githubusercontent.com/134745/67628636-2d104f00-f826-11e9-957d-a844861e4051.png)

The icon representing removal can be green, whereas keep can be red.

**Comment by westonruter** [comment]
Maybe it would make the most sense to just use the AMP logo? A red logo could be used for keep, whereas a green one for removal:

[![image](https://user-images.githubusercontent.com/134745/67628857-87f77580-f829-11e9-9c4e-00bba01d0adb.png)](https://codepen.io/westonruter/pen/ExxXEJJ)

When a URL has any kept validation errors, then the URL as a whole would get the “AMP Disabled” gray badge in Transitional mode, whereas the orange badge in Standard mode. If a URL has no kept validation errors, then it should get the “AMP Enabled” blue badge.

**Comment by amedina** [comment]
I like the ides of using AMP logos with color differentiation.

The only aspect I am not fully convinced is the cases for AMP disabled Standard vs. Transitional modes. It might be better to have both of those cases represented with the gray symbol.

**Comment by westonruter** [comment]
@amedina Yeah, going with gray will make it simpler; we won't have to explain it isn't _fully_ disabled.

Here's what it is looking like now:

<img width="1159" alt="Screen Shot 2019-10-27 at 18 07 09" src="https://user-images.githubusercontent.com/134745/67645236-b5edc000-f8e4-11e9-8225-9825b2aefaea.png">
<img width="1157" alt="Screen Shot 2019-10-27 at 18 07 33" src="https://user-images.githubusercontent.com/134745/67645237-b5edc000-f8e4-11e9-8626-362245d537a1.png">
<img width="1157" alt="Screen Shot 2019-10-27 at 18 07 47" src="https://user-images.githubusercontent.com/134745/67645238-b6865680-f8e4-11e9-8bd1-3244d4e3ec80.png">

**Comment by westonruter** [comment]
See all before/after screenshots in PR: https://github.com/ampproject/amp-wp/pull/3630

**Comment by csossi** [comment]
Verified in QA
